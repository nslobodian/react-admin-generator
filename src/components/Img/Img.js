import React, {Component} from 'react'
import pure from 'recompose/pure'

const magic = {
  minWidth: '100%',
  minHeight: '100%',
}

const IMAGE = 'image'
const COLOR = 'color'

export class Img extends Component {
  componentWillMount () {
    /* eslint-disable no-console */
    if (!this.props.children && !this.props.alt) {
      __DEV__ && console.error(`Image ${this.props.src} require alt prop`)
    }
    /* eslint-enable no-console */

    this.unmouted = false

    this.state = {
      loaded: false,
      error: false,
    }
  }

  componentDidMount = () => {
    this.loadNewImage(this.props.src)
  }

  componentWillUnmount = () => {
    this.unmouted = true
  }

  componentWillReceiveProps = ({src}) => {
    if (src !== this.props.src) {
      this.setState({
        loaded: false,
        error: false,
      })
      this.loadNewImage(src)
    }
  }

  loadNewImage = src => {
    new Promise((resolve, reject) => {
      let image = new Image()
      image.src = src

      image.onload = resolve
      image.onerror = reject
    }).then(() => {
      !this.unmouted && this.setState({loaded: true})
    }).catch(() => {
      !this.unmouted && this.setState({error: true})
    })
  }

  getWrapperStyle = () => {
    const {placeholderType, placeholder, src} = this.props
    const {loaded} = this.state

    switch (placeholderType) {
      case IMAGE:
        return {backgroundImage: `url(${loaded && src ? src : placeholder})`}

      case COLOR:
        return loaded && src
          ? {backgroundImage: `url(${src})`}
          : {backgroundColor: `rgba(${placeholder})`}

      default:
        return {}
    }
  }

  render () {
    const {
      src,
      placeholder,
      children,
      imageProps,
      placeholderProps,
      placeholderType,
      className,
      loadingClassName,
      alt,
    } = this.props
    const {loaded, error} = this.state

    return children
      ? <div
        style={this.getWrapperStyle()}
        className={loaded ? className : loadingClassName || className}
      >
        {children}
      </div>
      : loaded && !error && src
        ? <img
          src={src}
          alt={alt}
          className={className}
          {...imageProps}
        />
        : <Placeholder
          placeholderType={placeholderType}
          placeholder={placeholder}
          placeholderProps={placeholderProps}
          className={className}
        />
  }
}

export const Placeholder = pure(({placeholderType, placeholder, placeholderProps, className}) => {
  switch (placeholderType) {
    case IMAGE:
      return (
        <img
          src={placeholder}
          className={className}
          {...placeholderProps}
        />
      )

    case COLOR:
      return (
        <div
          style={{backgroundColor: `rgba(${placeholder})`, ...magic}}
          className={className}
          {...placeholderProps}
        />
      )
    default:
      return null
  }
})

export default pure(Img)
