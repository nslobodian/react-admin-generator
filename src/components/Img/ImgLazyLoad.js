import React from 'react'
import PropTypes from 'prop-types'
import LazyLoad from 'react-lazyload'
import Img from './Img'

const ImgLazyLoad = ({height = '100%', offset = 100, ...props}) =>
  <LazyLoad
    throttle={100}
    height={height}
    offset={offset}
    once
  >
    <Img {...props} />
  </LazyLoad>

ImgLazyLoad.propTypes = {
  props: PropTypes.object,
  height: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]),
  offset: PropTypes.number,
}
export default ImgLazyLoad
