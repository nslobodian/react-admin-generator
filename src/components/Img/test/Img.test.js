import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { Placeholder, Img } from '../Img'

describe('(Components) Placeholder', () => {
  it('Should render component properly with type image', () => {
    const placeholderType = 'image';
    const placeholder= 'placeholder';
    const placeholderProps = {};
    const className = 'placeholderClassName';
    const wrapper = shallow(
      <Placeholder
        placeholderType={placeholderType}
        placeholder={placeholder}
        placeholderProps={placeholderProps}
        className={className}
      />)
    expect(toJson(wrapper)).toMatchSnapshot()
  }),
  it('Should render component properly with type color', () => {
    const placeholderType = 'color';
    const placeholder= 'placeholder';
    const placeholderProps = {};
    const className = 'placeholderClassName';
    const wrapper = shallow(
      <Placeholder
        placeholderType={placeholderType}
        placeholder={placeholder}
        placeholderProps={placeholderProps}
        className={className}
      />)
    expect(toJson(wrapper)).toMatchSnapshot()
  }),
  it('Should render component properly with unmatching type', () => {
    const placeholderType = 'unmatchingType';
    const placeholder= 'placeholder';
    const placeholderProps = {};
    const className = 'placeholderClassName';
    const wrapper = shallow(
      <Placeholder
        placeholderType={placeholderType}
        placeholder={placeholder}
        placeholderProps={placeholderProps}
        className={className}
      />)
    expect(toJson(wrapper)).toMatchSnapshot()
  }),
  it('Should render component properly', () => {
    const props = {
      src: 'src',
      placeholder: 'placeholder',
      children: true,
      imageProps: {},
      placeholderProps: {},
      placeholderType: 'color',
      className: 'placeholderClassName',
      loadingClassName: true,
      loaded: true,
      alt: 'alterName',
    };
    const wrapper = shallow(
      <Img
        {...props}
      />)
    expect(toJson(wrapper)).toMatchSnapshot()
    wrapper.unmount()
  }),
  it('Should render component properly and will call didMount', () => {
    const props = {
      src: 'src',
      placeholder: 'placeholder',
      children: true,
      imageProps: {},
      placeholderProps: {},
      placeholderType: 'color',
      className: 'placeholderClassName',
      loadingClassName: true,
      alt: 'alterName',
    };
    const wrapper = shallow(
      <Img
        {...props}
      />)
    wrapper.instance().componentDidMount()
  }),
  it('Should render component properly and will call componentWillReceiveProps', () => {
    const props = {
      src: 'src',
      placeholder: 'placeholder',
      children: false,
      imageProps: {},
      placeholderProps: {},
      placeholderType: 'color',
      className: 'placeholderClassName',
      loadingClassName: true,
      loaded: true,
      alt: 'alterName',
    };
    const src = 'src';
    const wrapper = shallow(
      <Img
        {...props}
      />)
    wrapper.instance().componentWillReceiveProps({src})
  }),
  it('Should render component properly and will call componentWillReceiveProps with different src', () => {
    const props = {
      src: 'src',
      placeholder: 'placeholder',
      children: false,
      imageProps: {},
      placeholderProps: {},
      placeholderType: 'color',
      className: 'placeholderClassName',
      loadingClassName: true,
      loaded: true,
      alt: 'alterName',
    };
    const src = 'differentSrc';
    const wrapper = shallow(
      <Img
        {...props}
      />)
    wrapper.instance().componentWillReceiveProps({src})
  }),
  it('Should render component properly and will call loadNewImage', () => {
    const props = {
      src: 'src',
      placeholder: 'placeholder',
      children: false,
      imageProps: {},
      placeholderProps: {},
      placeholderType: 'color',
      className: 'placeholderClassName',
      loadingClassName: true,
      loaded: true,
      alt: 'alterName',
    };
    const src = 'differentSrc';
    const wrapper = shallow(
      <Img
        {...props}
      />)
    wrapper.instance().loadNewImage({src})
  }),
  it('Should render component properly and will call getWrapperStyle with placeholderType === color', () => {
    const props = {
      src: 'src',
      placeholder: 'placeholder',
      children: false,
      imageProps: {},
      placeholderProps: {},
      placeholderType: 'color',
      className: 'placeholderClassName',
      loadingClassName: true,
      loaded: true,
      alt: 'alterName',
    };
    const wrapper = shallow(
      <Img
        {...props}
      />)
    wrapper.instance().getWrapperStyle()
  }),
  it('Should render component properly and will call getWrapperStyle with placeholderType === image', () => {
    const props = {
      src: 'src',
      placeholder: 'placeholder',
      children: false,
      imageProps: {},
      placeholderProps: {},
      placeholderType: 'image',
      className: 'placeholderClassName',
      loadingClassName: true,
      loaded: true,
      alt: 'alterName',
    };
    const wrapper = shallow(
      <Img
        {...props}
      />)
    wrapper.instance().getWrapperStyle()
  }),
  it('Should render component properly and will call getWrapperStyle with unmatching placeholderType', () => {
    const props = {
      src: 'src',
      placeholder: 'placeholder',
      children: false,
      imageProps: {},
      placeholderProps: {},
      placeholderType: 'unmatchingPlaceholderType',
      className: 'placeholderClassName',
      loadingClassName: true,
      loaded: true,
      alt: 'alterName',
    };
    const wrapper = shallow(
      <Img
        {...props}
      />)
    wrapper.instance().getWrapperStyle()
  }),
  it('Should render component properly and will call componentWillMount', () => {
    const props = {
      src: 'src',
      placeholder: 'placeholder',
      children: false,
      imageProps: {},
      placeholderProps: {},
      placeholderType: 'unmatchingPlaceholderType',
      className: 'placeholderClassName',
      loadingClassName: true,
      loaded: true,
      alt: 'alterName',
    };
    const wrapper = shallow(
      <Img
        {...props}
      />)
    wrapper.instance().componentWillMount()
  })
});
