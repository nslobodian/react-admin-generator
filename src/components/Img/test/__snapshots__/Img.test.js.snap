// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Components) Placeholder Should render component properly 1`] = `
<div
  className={true}
  style={
    Object {
      "backgroundColor": "rgba(placeholder)",
    }
  }
/>
`;

exports[`(Components) Placeholder Should render component properly with type color 1`] = `
<div
  className="placeholderClassName"
  style={
    Object {
      "backgroundColor": "rgba(placeholder)",
      "minHeight": "100%",
      "minWidth": "100%",
    }
  }
/>
`;

exports[`(Components) Placeholder Should render component properly with type image 1`] = `
<img
  className="placeholderClassName"
  src="placeholder"
/>
`;

exports[`(Components) Placeholder Should render component properly with unmatching type 1`] = `null`;
